Implement explicit environment spec file support based on `CEP 23`.

jezdez opened the pull request:
<!-- Hello! Thanks for submitting a PR! To help make things go a bit more
     smoothly, we would appreciate it if you follow this template. -->

### Description

- Fix #14819.
- Fix https://github.com/conda/conda-environments-project-mgmt/issues/47

<!-- Good things to put here include:
       - reasons for the change (please link any relevant issues!),
       - any noteworthy (or hacky) choices to be aware of,
       - or what the problem resolved here looked like. -->

### Checklist - did you ...

<!-- If any of the following items aren't relevant to your contribution,
     please either tick them or use ~strikethrough~ so we know you've gone
     through the checklist. -->

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?
     <!-- All "significant" changes should get an entry:
            - user-facing changes or enhancements
            - bug fixes
            - deprecations
            - documentation updates
            - etc -->
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?


<!-- Just as a reminder, everyone in all conda org spaces (including PRs)
     must follow the Conda Org Code of Conduct (link below).

     Finally, once again, thanks for your time and effort. If you have any
     feedback in regards to your experience contributing here, please
     let us know!

     Helpful links:
       - Conda Org COC: https://github.com/conda/conda/blob/main/CODE_OF_CONDUCT.md
       - Contributing docs: https://github.com/conda/conda/blob/main/CONTRIBUTING.md -->

jezdez committed:
Implement explicit environment spec file support based on CEP-23.

jezdez committed:
Implement explicit environment spec file support based on CEP-23.

jezdez committed:
Ruff.

jezdez committed:
Ruff.

jezdez committed:
Fix year number.

jezdez committed:
Fix year number.

jezdez committed:
Fix year numbers because we code like it's 2012

jezdez committed:
Fix year numbers because we code like it's 2012

jjhelmus commented:
Creating an environment using this plugin requires a solve. A solve is not required when using `conda create --file explicit.txt`. Should this be documented or mentioned to users?

jjhelmus commented:
Testing this out it looks like all explicit files are detected as being handled by the `requirements.txt` plugin as well.

```
❯ cat explicit_single.txt 
@EXPLICIT
https://conda.anaconda.org/conda-forge/osx-64/libcurl-7.69.1-hc0b9707_0.tar.bz2

❯ conda env create --prefix blah --file explicit_single.txt

PluginError: Too many plugins found that can handle the environment file 'explicit_single.txt':

explicit, requirements.txt

Please make sure that you don't have any overlapping plugins installed.
```

jezdez committed:
Consolidate explicit environment handling into dedicated module

jezdez committed:
Consolidate explicit environment handling into dedicated module

jezdez committed:
Ruff.

jezdez committed:
Ruff.

jjhelmus commented on conda/env/installers/conda.py:
```diff
@@ -43,8 +59,35 @@ def dry_run(specs, args, env, *_, **kwargs):
+        This implementation follows CEP-23, which states: "When an explicit input file is
+        processed, the conda client SHOULD NOT invoke a solver."
+    """
+    # Handle explicit environments separately per CEP-23 requirements
+    if isinstance(env, ExplicitEnvironment):
+        return _install_explicit_environment(prefix, specs, env)
```
This is a special case but one that will come up for any plugins that reads a definition for a locked environment (including pixi.lock or conda-lock.yml files).  In these cases a solve is unnecessary, should not be performed and in some cases (e.g. packages outside of a channel are referenced) may not be possible. I think providing a path for plugins to express this information would be useful but likely requires an expansion of the `Environment` class.

For pixi.lock and conda-lock.yml, parsing can return a set of `PackageRecord` instances for the conda requirements, where as requirement and environment.yaml files would return a set of `MatchSpec` instances.

jjhelmus replied:
Another option would be to attaching this metadata to the `Environment` class itself which is returned from `EnvironmentSpecBase.environment` method. 

Either option is reasonable and will allow for plugins for various lock formats.

jjhelmus replied:
Is there a need to get this feature ASAP? If not I'd prefer to wait until the Environment class is merged. Once that work is done this can be re-worked to use that class.

soapy1 commented on conda/env/installers/conda.py:
```diff
@@ -53,16 +96,79 @@ def install(prefix, specs, args, env, *_, **kwargs):
     except (UnsatisfiableError, SystemExit) as exc:
-        # See this comment for 'allow_retry' details
-        # https://github.com/conda/conda/blob/b4592e9eb0/conda/cli/install.py#L417-L429
-        if not getattr(exc, "allow_retry", True):
```
what is the motivation for removing this retry bit?

soapy1 reviewed (commented):
Neat! This works when I (pull in the changes from #14725 and) try to reproduce the issue described in https://github.com/conda/conda/pull/14725#issuecomment-2867623756.

soapy1 commented:
> I wonder if we should merge those PRs together?

For me, its ok if they are seperate since they implementation doesn't overlap and they have 2 distinct goals. But, if you think it would be helpful to merge them, that sounds fine as well.

soapy1 committed:
Return transaction action groups (#14835)

soapy1 committed:
Return transaction action groups (#14835)

jjhelmus commented on conda/env/installers/conda.py:
```diff
@@ -2,21 +2,30 @@
-    """Solve the environment"""
+    """Solve the environment.
+
+    :param prefix: Installation target directory
+    :param specs: Package specifications to install
+    :param args: Command-line arguments
+    :param env: Environment object
+    :return: Solver object
```
Is it worthing adding or convert these to type annotation?

jjhelmus commented on conda/env/installers/conda.py:
```diff
@@ -53,16 +96,78 @@ def install(prefix, specs, args, env, *_, **kwargs):
+    explicit_specs = None
+    filename = env.explicit_filename
+
+    # 1. Try to read from original file if available (most reliable source)
+    if filename and os.path.exists(filename):
+        try:
+            with open(filename) as f:
+                explicit_specs = [
+                    line.strip()
+                    for line in f
+                    if line.strip() and not line.strip().startswith("#")
+                ]
+            log.debug(f"Using package specs from explicit file: {filename}")
+        except OSError as e:
+            log.warning(f"Could not read explicit file {filename}: {e}")
```
This logic exists in two (or even three see conda.cli.spec_from_line) locations. Can this be de-duped or collected into a utility function. 
Fine leaving this as a future refactor.

jjhelmus commented on conda/env/installers/conda.py:
```diff
@@ -53,16 +96,78 @@ def install(prefix, specs, args, env, *_, **kwargs):
+
+def _install_explicit_environment(
+    prefix: str, specs: list, env: ExplicitEnvironment
+) -> dict:
```
```suggestion
) -> dict[str, Any]:
```

jezdez committed:
Update conda/env/installers/conda.py

Co-authored-by: Jonathan J. Helmus <jjhelmus@gmail.com>

jezdez committed:
Update conda/env/installers/conda.py

Co-authored-by: Jonathan J. Helmus <jjhelmus@gmail.com>

jezdez committed:
Add type hints to conda installer functions and create explicit package list

jezdez committed:
Add type hints to conda installer functions and create explicit package list

jezdez committed:
Reset removed retry.

jezdez committed:
Reset removed retry.

jezdez committed:
refactor: add read_non_comment_lines utility and use it across file reading operations

jezdez committed:
refactor: add read_non_comment_lines utility and use it across file reading operations

jezdez committed:
adding missing future import

jezdez committed:
adding missing future import

jezdez committed:
Linting fixes.

jezdez committed:
Linting fixes.

soapy1 commented:
In general, I think this looks good to me and seems to work in all the ways I expect it to. I want to call out 2 other things happening in parallel here:

1) once the environment model stuff get's merged we'll want to change the environment spec plugins to be using that environment opposed to the `conda.env.env.Environment`. I think this change is fully compatible with what is happening in this PR. But merging this first will add to the work required to move over to that desired final state of all plugins using the new environment model.

2) @jjhelmus's [pr for implementing installing from a lockfile](https://github.com/conda/conda/pull/14868/files#diff-770280f144ccd2f54a49acb17077dd9109c45b8cd1fd8f26c594e728dbfad6c5) proposes another way of installing explicit packages. Is this the direction we want to go in for installing packages directly? If so, is this PR compatible with being able to move over to this new method?

soapy1 commented:
> I'm not sure I follow, are you asking if we can first merge this environment model and then this feature branch?

Sorry, that was a lot of words on my part. I think this is good to merge as is.

jezdez committed:
Add news.

jezdez committed:
Add news.

jezdez committed:
Update conda/env/installers/conda.py

Co-authored-by: Mahe Iram Khan <65779580+ForgottenProgramme@users.noreply.github.com>

jezdez committed:
Update conda/env/installers/conda.py

Co-authored-by: Mahe Iram Khan <65779580+ForgottenProgramme@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jaimergp reviewed (changes requested):
I have several concerns with this PR:

- I'm not sure I would consider `@EXPLICIT` a different format than `requirements.txt`, so I'm not sure I would have a different backend for this instead of teaching the existing `requirements` parser to handle explicit bits. After the only difference is what type of spec they ship (query vs resolved).
- Also, we are reimplementing behaviour existing in `conda create --file`, but with slightly different behavior (conversion of filepaths to URLs?); I don't recall the code doing that in `conda create`, but maybe I missed something. I'd strongly prefer if both commands end up using the very same logic.
- I'm not sure we want to introduce yet another `Environment` subclass, when we are talking about deprecating it in favor of https://github.com/conda/conda/pull/14870. That class will be able to handle both cases (explicit and not explicit) gracefully, so I don't know why we can't wait til that class ships before accepting this.


With all that in mind, , I thought we had a general goal of merging `conda env create` with `conda create` and end up deprecating `conda env create`. So I think it's more valuable to spend time teaching `conda create` how to do what `conda env` does, instead of the other way around. This is what @soapy1's refactors are directed to. Are we going to allow users to run `conda env create -f explicit.txt` in addition to `conda create --file explicit.txt` only to remove it later? I don't think it makes much sense to spend resources in this area, to be honest. We are adding more code to do the same thing in a different area of the CLI, adding tests and allocating engineering resources to something that will be eventually phased out.

If this was an exercise to try out the new plugin hooks, then cool with me, but I fail to see the added value if we accept that `conda env` will be eventually removed. 

There's a scenario where the conda env plugins end up backing up the logic behind `conda create`. In that case, this code might be useful in the future when we get there, but it feels like we should focus first on pluginizing `conda create` input file support after having gone through the necessary refactors.

jezdez committed:
Update conda/env/specs/requirements.py

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

jezdez committed:
Update conda/env/specs/requirements.py

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

jaimergp commented:
> Yep, when this PR was started we only barely started the discussion that may end in something like https://github.com/conda/conda/pull/14870, I've asked for review for a bit now.

Ok, I missed that temporal context, thanks for the reminder. That PR will get merged anytime now, so we can use it the moment it's in main.

> I'm sorry, but I need to clearly call out the unreasonable response to this, this PR doesn't stop unifying conda env create and conda create. I've discussed with Sophia that those refactors are worth having, but they haven't been merged yet.

Glad to hear that. For a moment I thought we were abandoning that adventure but good to see it's still on.

---

So to recap, I'll be happy to approve this once it uses Sophia's `Environment` data model. My other concerns about `conda create` vs `conda env create` are now gone with the replies above, since we will keep working towards that step by step. Thanks for baring with me!

jezdez committed:
Enhance URL handling in ExplicitRequirementsSpec

- Introduced `is_url` and `expand` functions for improved URL normalization and path expansion.
- Updated logic to handle file paths and URLs consistently, ensuring compatibility with existing specifications.
- Added validation for URL format using regex, maintaining backward compatibility while improving parsing accuracy.

jezdez committed:
Enhance URL handling in ExplicitRequirementsSpec

- Introduced `is_url` and `expand` functions for improved URL normalization and path expansion.
- Updated logic to handle file paths and URLs consistently, ensuring compatibility with existing specifications.
- Added validation for URL format using regex, maintaining backward compatibility while improving parsing accuracy.

jezdez committed:
Refactor explicit environment handling

- Removed the ExplicitEnvironment class and integrated its functionality into the Environment class.
- Updated the Dependencies class to include an explicit marker for environments created from @EXPLICIT files.
- Adjusted the conda installer and requirements specifications to work with the new Environment structure.
- Modified tests to reflect changes in the environment handling, ensuring compatibility with explicit specifications.

jezdez committed:
Refactor explicit environment handling

- Removed the ExplicitEnvironment class and integrated its functionality into the Environment class.
- Updated the Dependencies class to include an explicit marker for environments created from @EXPLICIT files.
- Adjusted the conda installer and requirements specifications to work with the new Environment structure.
- Modified tests to reflect changes in the environment handling, ensuring compatibility with explicit specifications.

jezdez committed:
Refactor requirements handling and unify explicit environment specifications

- Replaced the ExplicitRequirementsSpec with a unified RequirementsSpec that handles both regular and explicit requirements files.
- Updated the reading logic to utilize a generator for improved efficiency and clarity.
- Adjusted tests to reflect the new unified handling of requirements, ensuring compatibility with explicit specifications.
- Removed deprecated functions and streamlined the codebase for better maintainability.

jezdez committed:
Refactor requirements handling and unify explicit environment specifications

- Replaced the ExplicitRequirementsSpec with a unified RequirementsSpec that handles both regular and explicit requirements files.
- Updated the reading logic to utilize a generator for improved efficiency and clarity.
- Adjusted tests to reflect the new unified handling of requirements, ensuring compatibility with explicit specifications.
- Removed deprecated functions and streamlined the codebase for better maintainability.

jezdez committed:
Implement explicit environment installation logic in conda installer

- Integrated explicit environment handling directly into the install function, removing the need for a separate _install_explicit_environment function.
- Enhanced logging to provide detailed feedback on the source of package specifications used during installation.
- Improved the logic for determining package specs by prioritizing original files, provided specs, and environment dependencies.
- Ensured compatibility with the latest requirements handling and streamlined the installation process for explicit environments.

jezdez committed:
Implement explicit environment installation logic in conda installer

- Integrated explicit environment handling directly into the install function, removing the need for a separate _install_explicit_environment function.
- Enhanced logging to provide detailed feedback on the source of package specifications used during installation.
- Improved the logic for determining package specs by prioritizing original files, provided specs, and environment dependencies.
- Ensured compatibility with the latest requirements handling and streamlined the installation process for explicit environments.

jezdez committed:
Refactor transaction execution in conda installer

- Removed the _execute_transaction function and integrated its logic directly into the install function.
- Enhanced the transaction handling by checking for 'nothing to do' before executing.
- Streamlined the process of downloading, extracting, and executing transactions, improving clarity and maintainability.

jezdez committed:
Refactor transaction execution in conda installer

- Removed the _execute_transaction function and integrated its logic directly into the install function.
- Enhanced the transaction handling by checking for 'nothing to do' before executing.
- Streamlined the process of downloading, extracting, and executing transactions, improving clarity and maintainability.

jezdez committed:
Remove unused imports and clean up whitespace in conda installer

jezdez committed:
Remove unused imports and clean up whitespace in conda installer

jezdez committed:
Remove it back

jezdez committed:
Remove it back

jezdez committed:
Update conda/env/env.py

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

jezdez committed:
Update conda/env/env.py

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

jaimergp commented on conda/env/installers/conda.py:
```diff
@@ -29,22 +48,90 @@ def _solve(prefix, specs, args, env, *_, **kwargs):
     subdirs = IndexedSet(basename(url) for url in _channel_priority_map)
 
     solver_backend = context.plugin_manager.get_cached_solver_backend()
+    assert solver_backend is not None  # get_cached_solver_backend never returns None
```
Should we raise a `CondaError` subclass here?

jaimergp commented on conda/env/installers/conda.py:
```diff
@@ -29,22 +48,90 @@ def _solve(prefix, specs, args, env, *_, **kwargs):
+       package set before installation.
+
+    :param prefix: The target installation path for the environment
+    :type prefix: str
```
I don't think we need to double annotate the types here if we have type hints already.

jaimergp commented on conda/env/installers/conda.py:
```diff
@@ -29,22 +48,90 @@ def _solve(prefix, specs, args, env, *_, **kwargs):
+) -> dict | None:
+    """Install packages into a conda environment.
+
+    This function handles two main paths:
+    1. For environments with explicit_specs (from @EXPLICIT files): Bypasses the solver
+       and directly installs packages using conda.misc.explicit() as required by CEP-23.
+    2. For regular Environment instances: Uses the solver to determine the optimal
+       package set before installation.
```
Love this bit! Super clear :)

jaimergp commented on conda/env/installers/conda.py:
```diff
@@ -29,22 +48,90 @@ def _solve(prefix, specs, args, env, *_, **kwargs):
+        if filename:
+            explicit_specs = list(yield_lines(filename))
+            if explicit_specs:
+                log.debug(f"Using package specs from explicit file: {filename}")
```
```suggestion
                log.debug("Using package specs from explicit file: %s", filename})
```

jaimergp commented on conda/env/installers/conda.py:
```diff
@@ -29,22 +48,90 @@ def _solve(prefix, specs, args, env, *_, **kwargs):
+            if explicit_specs:
+                log.debug(f"Using package specs from explicit file: {filename}")
+            else:
+                log.warning(f"Could not read explicit file {filename} or file is empty")
```
```suggestion
                log.warning("Could not read explicit file %s or file is empty", filename)
```

jaimergp commented on conda/gateways/disk/read.py:
```diff
@@ -50,12 +50,17 @@ def yield_lines(path):
     try:
-        with open_utf8(path) as fh:
-            for line in fh:
+        # Convert to Path object for consistent handling
```
I think this comment is a bit superfluous and we don't need it.

jaimergp commented on tests/env/test_explicit_environment.py:
```diff
@@ -0,0 +1,89 @@
+
+def test_explicit_environment_initialization(explicit_urls):
+    """Test Environment initialization with explicit parameters."""
+    explicit_deps = ["@EXPLICIT"] + explicit_urls
```
Can we add a test where `@EXPLICIT` is not the first line, just to make sure it works elsewhere in the file?

jaimergp commented on conda/env/installers/conda.py:
```diff
@@ -29,22 +48,90 @@ def _solve(prefix, specs, args, env, *_, **kwargs):
+                log.warning(f"Could not read explicit file {filename} or file is empty")
+
+        # 2-3. Fall back to provided specs or dependencies from environment
+        if not explicit_specs:
+            explicit_specs = specs if specs else env.dependencies.raw
+            log_msg = (
+                "Using specs provided to installer"
+                if specs
+                else "Using dependencies from environment"
+            )
+            log.debug(log_msg)
```
I don't fully understand the purpose of this fallback. If the explicit_dependencies are empty, there's nothing to install.

What we are missing here is to use the non-explicit specs as the `history` contents, but I think that's missing in the current `explicit()` function API anyway.

jaimergp reviewed (commented):
I have some questions about the precedence of the `explicit()` code paths in the absence of `explicit_specs`. Other than that, the rest looks solid. I'll take a look locally tomorrow and extend my review with maybe some more comments. In the meantime we can address the non controversial feedback. Thanks!

jezdez committed:
Replace assertion with a conditional check that raises a CondaValueError if no solver backend is found.

jezdez committed:
Replace assertion with a conditional check that raises a CondaValueError if no solver backend is found.

jezdez committed:
- Updated logging statements in `conda.py` to use parameterized logging for better performance and readability.
- Removed unnecessary comments and improved the `yield_lines` function in `read.py` for cleaner code.

jezdez committed:
- Updated logging statements in `conda.py` to use parameterized logging for better performance and readability.
- Removed unnecessary comments and improved the `yield_lines` function in `read.py` for cleaner code.

jezdez committed:
Add tests for handling @EXPLICIT marker in dependencies

- Introduced `test_explicit_marker_in_middle` to verify functionality of the @EXPLICIT marker when placed in the middle of the dependency list.
- Added `test_explicit_marker_at_end` to ensure the @EXPLICIT marker works correctly when positioned at the end of the dependencies.

jezdez committed:
Add tests for handling @EXPLICIT marker in dependencies

- Introduced `test_explicit_marker_in_middle` to verify functionality of the @EXPLICIT marker when placed in the middle of the dependency list.
- Added `test_explicit_marker_at_end` to ensure the @EXPLICIT marker works correctly when positioned at the end of the dependencies.

jezdez committed:
Improve explicit function handling and enhance type hinting

- Updated the `explicit` function to track user-requested specs separately from all packages, preventing dependencies from polluting the user's request history.
- Added type hints for parameters in `explicit`, `_solve`, `dry_run`, and `install` functions to improve code clarity and type checking.
- Modified test cases to ensure correct handling of the `@EXPLICIT` marker in various positions within the dependency list.

jezdez committed:
Improve explicit function handling and enhance type hinting

- Updated the `explicit` function to track user-requested specs separately from all packages, preventing dependencies from polluting the user's request history.
- Added type hints for parameters in `explicit`, `_solve`, `dry_run`, and `install` functions to improve code clarity and type checking.
- Modified test cases to ensure correct handling of the `@EXPLICIT` marker in various positions within the dependency list.

jezdez committed:
Refactor explicit environment tests and enhance functionality

- Removed redundant fixture and consolidated tests for explicit environments.
- Added new tests to verify the detection of explicit dependencies and the handling of the @EXPLICIT marker in various positions within the dependency list.
- Improved type hinting for test functions and ensured consistent assertions for environment properties.

jezdez committed:
Refactor explicit environment tests and enhance functionality

- Removed redundant fixture and consolidated tests for explicit environments.
- Added new tests to verify the detection of explicit dependencies and the handling of the @EXPLICIT marker in various positions within the dependency list.
- Improved type hinting for test functions and ensured consistent assertions for environment properties.

jezdez committed:
Refactor assertions in explicit environment tests for clarity

- Updated assertions in `test_explicit_file_spec_is_registered` to improve readability and ensure correct handling of dependencies.
- Replaced direct access to `env.dependencies["conda"]` with a more general approach using `env.dependencies.raw`, enhancing flexibility in dependency checks.

jezdez committed:
Refactor assertions in explicit environment tests for clarity

- Updated assertions in `test_explicit_file_spec_is_registered` to improve readability and ensure correct handling of dependencies.
- Replaced direct access to `env.dependencies["conda"]` with a more general approach using `env.dependencies.raw`, enhancing flexibility in dependency checks.

soapy1 committed:
Fix tests

jezdez committed:
Merge remote-tracking branch 'soapy1/explicit-env-spec-2' into explicit-env-spec

jezdez committed:
Updated the `update_specs_for_history` logic to ensure only user-requested specs are recorded, preventing dependencies from affecting the user's request history.
Also enhanced the handling of requested specs by using `MatchSpec` for better specification matching.

jezdez committed:
Fix comment.
